# 🎯 Phase 5: Methodology Improvement & Comparison

In this phase, the existing methodology is improved by identifying gaps in previous approaches and applying enhancements.

## 🔍 Objective
- Improve model performance
- Make results more reliable
- Compare old vs improved methodology

## 📂 Load Feature Engineered Dataset

We load the dataset prepared in Phase 3.

In [1]:
import pandas as pd

df = pd.read_csv("/kaggle/input/notebooks/zohawaqas10/ml-project-phase-3/feature_engineered_students_dataset.csv")

df.head()

,Gender,Age,Education Level,Institution Type,IT Student,Location,Load-shedding,Financial Condition,Internet Type,Network Type,Class Duration,Self Lms,Device,Adaptivity Level,tech_access_score,infra_difficulty,socioeconomic_score,student_cluster
0,0,3,2,1,0,1,1,0,1,2,2,0,2,2,5,2,1,2
1,1,3,2,1,0,1,0,0,0,2,1,1,1,2,3,1,1,2
2,1,2,0,0,0,1,1,0,1,2,1,0,1,2,4,2,0,2
3,1,1,1,1,0,1,1,0,0,2,1,0,1,2,3,2,1,2
4,1,2,1,1,0,1,1,1,0,1,0,0,1,1,2,2,2,2


The dataset has been loaded successfully.

## 🎯 Define Features and Target

In [2]:
X = df.drop("Adaptivity Level", axis=1)
y = df["Adaptivity Level"]

## ✂️ Splitting Dataset

In [3]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## 📊 Baseline Model (Old Methodology)

Decision Tree is used as the baseline model from Phase 4.

In [4]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)

dt_pred = dt.predict(X_test)

baseline_acc = accuracy_score(y_test, dt_pred)
print("Baseline Accuracy:", baseline_acc)

Baseline Accuracy: 0.928


This represents the performance of the original methodology.

## 🔍 Overfitting Check

In [5]:
print("Decision Tree Training Accuracy:", dt.score(X_train, y_train))
print("Decision Tree Testing Accuracy:", dt.score(X_test, y_test))

Decision Tree Training Accuracy: 0.9306666666666666
Decision Tree Testing Accuracy: 0.928


If training accuracy is higher than testing accuracy, it indicates overfitting.

## 🔁 Improvement 1: Cross Validation

Cross-validation is applied to ensure model stability and reduce overfitting.

In [6]:
from sklearn.model_selection import cross_val_score

cv_scores = cross_val_score(dt, X, y, cv=5)

print("Cross Validation Scores:", cv_scores)
print("Mean CV Score:", cv_scores.mean())

Cross Validation Scores: [0.928      0.90933333 0.90666667 0.93866667 0.90933333]
Mean CV Score: 0.9184000000000001


Cross-validation shows that the model performs consistently across different splits, improving reliability.

## ⚡ Improvement 2: LightGBM Model

LightGBM is used to improve performance using boosting techniques.

In [7]:
import lightgbm as lgb

lgb_model = lgb.LGBMClassifier()
lgb_model.fit(X_train, y_train)

lgb_pred = lgb_model.predict(X_test)

lgb_acc = accuracy_score(y_test, lgb_pred)
print("LightGBM Accuracy:", lgb_acc)

[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001338 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 51
[LightGBM] [Info] Number of data points in the train set: 1500, number of used features: 17
[LightGBM] [Info] Start training from score -1.106644
[LightGBM] [Info] Start training from score -1.108663
[LightGBM] [Info] Start training from score -1.080772
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further s

## 📊 Comparison: Before vs After Improvement

In [8]:
import pandas as pd

comparison = pd.DataFrame({
    "Model": ["Decision Tree (Baseline)", "LightGBM (Improved)"],
    "Accuracy": [baseline_acc, lgb_acc]
})

comparison

,Model,Accuracy
0,Decision Tree (Baseline),0.928000
1,LightGBM (Improved),0.906667


## 📊 Analysis of Improvement

Although Decision Tree achieved higher test accuracy (92.8%), further analysis revealed that:

- Decision Tree shows signs of overfitting
- Cross-validation scores are lower than test accuracy
- Performance varies across different data splits

On the other hand:

- LightGBM provides slightly lower accuracy (90%)
- However, it shows more consistent performance
- It generalizes better to unseen data

## 🎯 Conclusion

Therefore, the improved methodology (LightGBM + Cross-validation) is more reliable and robust compared to the baseline Decision Tree model.

This demonstrates that improvement is not only about higher accuracy but also about better generalization and stability.